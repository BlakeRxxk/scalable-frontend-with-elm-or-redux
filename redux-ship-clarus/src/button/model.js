// @flow

export type State = {
  status: 'green' | 'red',
};

export const initialState: State = {
  status: 'green',
};

export type Patch = {
  type: 'Toggle',
};

export function reduce(state: State, patch: Patch): State {
  switch (patch.type) {
  case 'Toggle':
    return {
      ...state,
      status: state.status === 'green' ? 'red' : 'green',
    };
  default:
    return state;
  }
}
