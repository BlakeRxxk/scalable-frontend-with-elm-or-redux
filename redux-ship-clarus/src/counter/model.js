// @flow

export type State = {
  count: number,
};

export const initialState: State = {
  count: 0,
};

export type Action = {
  type: 'IncrementByOne',
} | {
  type: 'IncrementByTwo',
};

export function reduce(state: State, action: Action): State {
  switch (action.type) {
  case 'IncrementByOne':
    return {
      ...state,
      count: state.count + 1,
    };
  case 'IncrementByTwo':
    return {
      ...state,
      count: state.count + 2,
    };
  default:
    return state;
  }
}
