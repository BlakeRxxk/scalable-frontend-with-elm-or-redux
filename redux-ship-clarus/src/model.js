// @flow
import * as ButtonModel from './button/model';
import * as CounterModel from './counter/model';
import * as RandomGifModel from './random-gif/model';
import * as RandomGifPairModel from './random-gif-pair/model';

export type State = {
  button: ButtonModel.State,
  counter: CounterModel.State,
  randomGif: RandomGifModel.State,
  randomGifPair: RandomGifPairModel.State,
};

export const initialState: State = {
  button: ButtonModel.initialState,
  counter: CounterModel.initialState,
  randomGif: RandomGifModel.initialState,
  randomGifPair: RandomGifPairModel.initialState,
};

export type Patch = {
  button?: ButtonModel.Patch,
  counter?: CounterModel.Patch,
  randomGif?: RandomGifModel.Patch,
  randomGifPair?: RandomGifPairModel.Patch,
};

export function reduce(state: State, patch: Patch): State {
  return {
    ...state,
    ...patch.button &&
      {button: ButtonModel.reduce(state.button, patch.button)},
    ...patch.counter &&
      {counter: CounterModel.reduce(state.counter, patch.counter)},
    ...patch.randomGif &&
      {randomGif: RandomGifModel.reduce(state.randomGif, patch.randomGif)},
    ...patch.randomGifPair &&
      {randomGifPair: RandomGifPairModel.reduce(state.randomGifPair, patch.randomGifPair)},
  };
}
