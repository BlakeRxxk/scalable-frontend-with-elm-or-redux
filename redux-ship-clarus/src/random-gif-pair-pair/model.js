// @flow
import * as RandomGifPairModel from '../random-gif-pair/model';

export type State = {
  first: RandomGifPairModel.State,
  second: RandomGifPairModel.State,
};

export const initialState: State = {
  first: RandomGifPairModel.initialState,
  second: RandomGifPairModel.initialState,
};

export type Patch = {
  first?: RandomGifPairModel.Patch,
  second?: RandomGifPairModel.Patch,
};

export function reduce(state: State, patch: Patch): State {
  return {
    ...state,
    ...patch.first && {first: RandomGifPairModel.reduce(state.first, patch.first)},
    ...patch.second && {second: RandomGifPairModel.reduce(state.second, patch.second)},
  };
}
