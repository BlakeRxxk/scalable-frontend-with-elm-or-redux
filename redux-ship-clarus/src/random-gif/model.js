// @flow

export type State = {
  gifUrl: ?string,
  isLoading: bool,
};

export const initialState: State = {
  gifUrl: null,
  isLoading: false,
};

export type Action = {
  type: 'LoadStart',
} | {
  type: 'LoadSuccess',
  gifUrl: string,
};

export function reduce(state: State, action: Action): State {
  switch (action.type) {
  case 'LoadStart':
    return {
      ...state,
      isLoading: true,
    };
  case 'LoadSuccess':
    return {
      ...state,
      isLoading: false,
      gifUrl: action.gifUrl,
    };
  default:
    return state;
  }
}
