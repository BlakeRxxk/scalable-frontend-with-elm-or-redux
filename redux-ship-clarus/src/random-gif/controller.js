// @flow
import * as Ship from 'redux-ship';
import * as Effect from '../effect';
import * as CounterModel from '../counter/model';
import * as Model from './model';

export type Action = {
  type: 'Load',
  tag: string,
};

export type ModelAction = {
  type: 'RandomGif',
  action: Model.Action,
} | {
  type: 'Counter',
  action: CounterModel.Action,
};

export type ModelState = {
  counter: CounterModel.State,
  randomGif: Model.State,
};

function* incrementCounter() {
  const {count} = (yield* Ship.getState()).counter;
  if (count >= 10) {
    yield* Ship.dispatch({
      type: 'Counter',
      action: {
        type: 'IncrementByTwo',
      },
    });
  } else {
    yield* Ship.dispatch({
      type: 'Counter',
      action: {
        type: 'IncrementByOne',
      },
    });
  }
}

export function* control(
  action: Action
): Ship.Ship<Effect.Effect, ModelAction, ModelState, void> {
  switch (action.type) {
  case 'Load': {
    yield* Ship.dispatch({
      type: 'RandomGif',
      action: {
        type: 'LoadStart',
      },
    });
    const result = yield* Effect.httpRequest(
      `http://api.giphy.com/v1/gifs/random?api_key=dc6zaTOxFJmzC&tag=${action.tag}`
    );
    const gifUrl: string = JSON.parse(result).data.image_url;
    yield* Ship.dispatch({
      type: 'RandomGif',
      action: {
        type: 'LoadSuccess',
        gifUrl,
      },
    });
    yield* incrementCounter();
    return;
  }
  default:
    return;
  }
}
