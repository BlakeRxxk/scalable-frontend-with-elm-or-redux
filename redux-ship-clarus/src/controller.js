// @flow
import * as Ship from 'redux-ship';
import * as Effect from './effect';
import * as RandomGifController from './random-gif/controller';
import * as RandomGifPairController from './random-gif-pair/controller';
import * as Model from './model';

export type Action = {
  type: 'RandomGif',
  action: RandomGifController.Action,
} | {
  type: 'RandomGifPair',
  action: RandomGifPairController.Action,
};

export function* control(
  action: Action
): Ship.Ship<Effect.Effect, Model.Action, Model.State, void> {
  switch (action.type) {
  case 'RandomGif': {
    return yield* Ship.map(
      action => ({type: 'RandomGif', action}),
      state => state.randomGif,
      RandomGifController.control(action.action)
    );
  }
  case 'RandomGifPair': {
    return yield* Ship.map(
      action => ({type: 'RandomGifPair', action}),
      state => state.randomGifPair,
      RandomGifPairController.control(action.action)
    );
  }
  default:
    return;
  }
}
