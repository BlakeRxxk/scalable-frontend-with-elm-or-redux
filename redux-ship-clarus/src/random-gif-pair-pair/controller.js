// @flow
import * as Ship from 'redux-ship';
import * as ButtonModel from '../button/model';
import * as CounterModel from '../counter/model';
import * as RandomGifPairController from '../random-gif-pair/controller';
import * as RandomGifPairPairModel from './model';

export type Action = {
  type: 'First',
  action: RandomGifPairController.Action,
} | {
  type: 'Second',
  action: RandomGifPairController.Action,
};

export type Commit = {
  type: 'First',
  commit: RandomGifPairController.Commit,
} | {
  type: 'Second',
  commit: RandomGifPairController.Commit,
};

export type State = {
  button: ButtonModel.State,
  counter: CounterModel.State,
  randomGifPairPair: RandomGifPairPairModel.State,
};

export type Patch = {
  counter?: CounterModel.Patch,
  randomGifPairPair?: RandomGifPairPairModel.Patch,
};

export function applyCommit(state: State, commit: Commit): Patch {
  switch (commit.type) {
  case 'First': {
    const patch = RandomGifPairController.applyCommit(
      {
        button: state.button,
        counter: state.counter,
        randomGifPair: state.randomGifPairPair.first
      },
      commit.commit
    );
    return {
      ...patch.counter && {counter: patch.counter},
      ...patch.randomGifPair && {randomGifPairPair: {first: patch.randomGifPair}},
    };
  }
  case 'Second': {
    const patch = RandomGifPairController.applyCommit(
      {
        button: state.button,
        counter: state.counter,
        randomGifPair: state.randomGifPairPair.second
      },
      commit.commit
    );
    return {
      ...patch.counter && {counter: patch.counter},
      ...patch.randomGifPair && {randomGifPairPair: {second: patch.randomGifPair}},
    };
  }
  default:
    return {};
  }
}

export function* control(action: Action): Ship.Ship<*, Commit, State, void> {
  switch (action.type) {
  case 'First':
    return yield* Ship.map(
      commit => ({type: 'First', commit}),
      state => ({
        button: state.button,
        counter: state.counter,
        randomGifPair: state.randomGifPairPair.first,
      }),
      RandomGifPairController.control(action.action)
    );
  case 'Second':
    return yield* Ship.map(
      commit => ({type: 'Second', commit}),
      state => ({
        button: state.button,
        counter: state.counter,
        randomGifPair: state.randomGifPairPair.second,
      }),
      RandomGifPairController.control(action.action)
    );
  default:
    return;
  }
}
