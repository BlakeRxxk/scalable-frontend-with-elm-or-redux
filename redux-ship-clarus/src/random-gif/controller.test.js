// @flow
import * as Test from '../test';
import * as RandomGifController from './controller';
import * as ButtonModel from '../button/model';
import * as CounterModel from '../counter/model';
import * as RandomGifModel from './model';

const defaultState = {
  button: ButtonModel.initialState,
  counter: CounterModel.initialState,
  randomGif: RandomGifModel.initialState,
};

Test.snapshotApplyCommit(RandomGifController.applyCommit, {
  LoadStart: {
    commit: {type: 'LoadStart'},
    state: defaultState,
  },
  LoadSuccessIncrementByOne: {
    commit: {
      type: 'LoadSuccess',
      gifUrl: 'https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif',
    },
    state: defaultState,
  },
  LoadSuccessIncrementByTwo: {
    commit: {
      type: 'LoadSuccess',
      gifUrl: 'https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif',
    },
    state: {
      ...defaultState,
      counter: {
        ...defaultState.counter,
        count: 10,
      },
    },
  },
});
