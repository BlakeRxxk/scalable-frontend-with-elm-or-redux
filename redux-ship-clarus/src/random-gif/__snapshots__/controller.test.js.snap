exports[`test LoadStart 1`] = `
Object {
  "randomGif": Object {
    "type": "LoadStart"
  }
}
`;

exports[`test LoadSuccessIncrementByOne 1`] = `
Object {
  "counter": Object {
    "type": "IncrementByOne"
  },
  "randomGif": Object {
    "gifUrl": "https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif",
    "type": "LoadSuccess"
  }
}
`;

exports[`test LoadSuccessIncrementByTwo 1`] = `
Object {
  "counter": Object {
    "type": "IncrementByTwo"
  },
  "randomGif": Object {
    "gifUrl": "https://media2.giphy.com/media/m7ychnf9zOVm8/giphy.gif",
    "type": "LoadSuccess"
  }
}
`;
